### ISSUE TITLE
[minillm] apply minillm to other LLMs like Baichuan/Qianwen

### ISSUE THREAD

**Issue by chentianyangWHU** [microsoft/LMOps#93]
Thank you for your awesome work minillm that explored the knowledge distillation for LLMs. I noticed that minillm supports the gpt2/gptj/opt and llama series models only, my question is how should I do if I want to extend it to more recently released LLMs like Baichuan-7B and Qwen-7B whose model architectures are different from llama/opt/gpt2?
I noticed that for big student/teacher models , it's necessary to split them into 4 parts to fit the A100 GPUs, so you provide folders like:
transformers/src/transformers/models/opt_parallel/
transformers/src/transformers/models/gpt2_parallel/
transformers/src/transformers/models/llama_parallel/
with necessary files inside, mainly 'modeling_xxx_parallel.py' and 'utils_xxx.py', to implement model split and parallel modeling,
However, when I checked the latest huggingface/transformers repository and didn't find those 'xxx_parallel/' folders. Is that mean those 'xxx_parallel/' folders and files inside are developed by yourself? And if I want to extend minillm to Baichuan or Qwen, do I need to development corresponding codes by myself, like, let's say 'baichuan_parallel/' and 'qwen_parallel/'?
Thank you!

**Comment by t1101675** [comment]
Yes, we implement the TensorParallel technique by ourselves, which is not supported by HuggingFace yet. You need to develop the `xxx_parallel/` folders by yourself. 
The development is based on Megatron-LM ([paper](https://arxiv.org/abs/1909.08053), [code](https://github.com/NVIDIA/Megatron-LM)) and basically includes three steps:
1. Replace the Linear layers in the Attention and FFN layers with the Parallel Linear layers defined in [mpu](https://github.com/microsoft/LMOps/blob/main/minillm/transformers/src/transformers/mpu/layers.py).
2. Replace the word embedding layers and the lm_head layers with the corresponding parallel implement in [mpu](https://github.com/microsoft/LMOps/blob/main/minillm/transformers/src/transformers/mpu/layers.py)
3. Replace the cross entropy loss calculation in the training scripts with the parallel version in [mpu](https://github.com/microsoft/LMOps/blob/main/minillm/transformers/src/transformers/mpu/cross_entropy.py)

I guess the model architecture of Baichuan-7B and Qwen-7B would not differ much from LLaMA, therefore, the development may not be quite difficult.

Alternatively, you can try other techniques if you find the GPU memory insufficient, like [LoRA](https://github.com/huggingface/peft) and [zero/cpu-offload](https://huggingface.co/docs/transformers/main_classes/deepspeed), which are already supported by HuggingFace or DeepSpeed.

**Comment by chentianyangWHU** [comment]
> Yes, we implement the TensorParallel technique by ourselves, which is not supported by HuggingFace yet. You need to develop the `xxx_parallel/` folders by yourself. The development is based on Megatron-LM ([paper](https://arxiv.org/abs/1909.08053), [code](https://github.com/NVIDIA/Megatron-LM)) and basically includes three steps:
> 
> 1. Replace the Linear layers in the Attention and FFN layers with the Parallel Linear layers defined in [mpu](https://github.com/microsoft/LMOps/blob/main/minillm/transformers/src/transformers/mpu/layers.py).
> 2. Replace the word embedding layers and the lm_head layers with the corresponding parallel implement in [mpu](https://github.com/microsoft/LMOps/blob/main/minillm/transformers/src/transformers/mpu/layers.py)
> 3. Replace the cross entropy loss calculation in the training scripts with the parallel version in [mpu](https://github.com/microsoft/LMOps/blob/main/minillm/transformers/src/transformers/mpu/cross_entropy.py)
> 
> I guess the model architecture of Baichuan-7B and Qwen-7B would not differ much from LLaMA, therefore, the development may not be quite difficult.
> 
> Alternatively, you can try other techniques if you find the GPU memory insufficient, like [LoRA](https://github.com/huggingface/peft) and [zero/cpu-offload](https://huggingface.co/docs/transformers/main_classes/deepspeed), which are already supported by HuggingFace or DeepSpeed.

Thank you for your patient answer! I am trying to incorporate baichuan and qwen into the minllm algorithm based on the clues and suggestions you provided.

**Comment by donglixp** [comment]
> > Yes, we implement the TensorParallel technique by ourselves, which is not supported by HuggingFace yet. You need to develop the `xxx_parallel/` folders by yourself. The development is based on Megatron-LM ([paper](https://arxiv.org/abs/1909.08053), [code](https://github.com/NVIDIA/Megatron-LM)) and basically includes three steps:
> > 
> > 1. Replace the Linear layers in the Attention and FFN layers with the Parallel Linear layers defined in [mpu](https://github.com/microsoft/LMOps/blob/main/minillm/transformers/src/transformers/mpu/layers.py).
> > 2. Replace the word embedding layers and the lm_head layers with the corresponding parallel implement in [mpu](https://github.com/microsoft/LMOps/blob/main/minillm/transformers/src/transformers/mpu/layers.py)
> > 3. Replace the cross entropy loss calculation in the training scripts with the parallel version in [mpu](https://github.com/microsoft/LMOps/blob/main/minillm/transformers/src/transformers/mpu/cross_entropy.py)
> > 
> > I guess the model architecture of Baichuan-7B and Qwen-7B would not differ much from LLaMA, therefore, the development may not be quite difficult.
> > Alternatively, you can try other techniques if you find the GPU memory insufficient, like [LoRA](https://github.com/huggingface/peft) and [zero/cpu-offload](https://huggingface.co/docs/transformers/main_classes/deepspeed), which are already supported by HuggingFace or DeepSpeed.
> 
> Thank you for your patient answer! I am trying to incorporate baichuan and qwen into the minllm algorithm based on the clues and suggestions you provided.

@chentianyangWHU  Contributions are welcome!

**Comment by chentianyangWHU** [comment]
> > > Yes, we implement the TensorParallel technique by ourselves, which is not supported by HuggingFace yet. You need to develop the `xxx_parallel/` folders by yourself. The development is based on Megatron-LM ([paper](https://arxiv.org/abs/1909.08053), [code](https://github.com/NVIDIA/Megatron-LM)) and basically includes three steps:
> > > 
> > > 1. Replace the Linear layers in the Attention and FFN layers with the Parallel Linear layers defined in [mpu](https://github.com/microsoft/LMOps/blob/main/minillm/transformers/src/transformers/mpu/layers.py).
> > > 2. Replace the word embedding layers and the lm_head layers with the corresponding parallel implement in [mpu](https://github.com/microsoft/LMOps/blob/main/minillm/transformers/src/transformers/mpu/layers.py)
> > > 3. Replace the cross entropy loss calculation in the training scripts with the parallel version in [mpu](https://github.com/microsoft/LMOps/blob/main/minillm/transformers/src/transformers/mpu/cross_entropy.py)
> > > 
> > > I guess the model architecture of Baichuan-7B and Qwen-7B would not differ much from LLaMA, therefore, the development may not be quite difficult.
> > > Alternatively, you can try other techniques if you find the GPU memory insufficient, like [LoRA](https://github.com/huggingface/peft) and [zero/cpu-offload](https://huggingface.co/docs/transformers/main_classes/deepspeed), which are already supported by HuggingFace or DeepSpeed.
> > 
> > 
> > Thank you for your patient answer! I am trying to incorporate baichuan and qwen into the minllm algorithm based on the clues and suggestions you provided.
> 
> @chentianyangWHU Contributions are welcome!

I am working on it. Coming soon. Thank you!

**Comment by WenTingTseng** [comment]
Looking forward to your launch.

**Comment by lllyyyqqq** [comment]
> Yes, we implement the TensorParallel technique by ourselves, which is not supported by HuggingFace yet. You need to develop the `xxx_parallel/` folders by yourself. The development is based on Megatron-LM ([paper](https://arxiv.org/abs/1909.08053), [code](https://github.com/NVIDIA/Megatron-LM)) and basically includes three steps:
> 
> 1. Replace the Linear layers in the Attention and FFN layers with the Parallel Linear layers defined in [mpu](https://github.com/microsoft/LMOps/blob/main/minillm/transformers/src/transformers/mpu/layers.py).
> 2. Replace the word embedding layers and the lm_head layers with the corresponding parallel implement in [mpu](https://github.com/microsoft/LMOps/blob/main/minillm/transformers/src/transformers/mpu/layers.py)
> 3. Replace the cross entropy loss calculation in the training scripts with the parallel version in [mpu](https://github.com/microsoft/LMOps/blob/main/minillm/transformers/src/transformers/mpu/cross_entropy.py)
> 
> I guess the model architecture of Baichuan-7B and Qwen-7B would not differ much from LLaMA, therefore, the development may not be quite difficult.
> 
> Alternatively, you can try other techniques if you find the GPU memory insufficient, like [LoRA](https://github.com/huggingface/peft) and [zero/cpu-offload](https://huggingface.co/docs/transformers/main_classes/deepspeed), which are already supported by HuggingFace or DeepSpeed.

Is Encoder-Decoder model like flan-t5 can be implemented TP like this ?
